src/audit/secrets_outside_environment.rs

```diff
@@ -0,0 +1,83 @@
+use github_actions_models::{
+    common::{Env, EnvValue, expr::LoE},
+    workflow::job::StepBody,
+};
+
+use super::{Audit, audit_meta};
+use crate::{finding::Confidence, models::StepCommon, AuditState, AuditLoadError, Persona};
+
+pub(crate) struct SecretsOutsideEnvironment;
+
+audit_meta!(
+    SecretsOutsideEnvironment,
+    "secrets-outside-environment",
+    "secrets used without an environment to gate them"
+);
+
+impl Audit for SecretsOutsideEnvironment {
+    fn new(_state: &AuditState<'_>) -> Result<Self, AuditLoadError>
+    where
+        Self: Sized,
+    {
+        Ok(Self)
+    }
+
+    fn audit_step<'w>(
+        &self,
+        step: &crate::models::Step<'w>,
+    ) -> anyhow::Result<Vec<crate::finding::Finding<'w>>> {
+        let mut findings = vec![];
+
+        if step.parent.environment().is_some() {
+            return Ok(findings);
+        }
+
+        let eenv: &Env;
+        match &step.body {
+            StepBody::Uses { uses: _, with } => {
+                eenv = with;
+            }
+            StepBody::Run {
+                run: _,
+                shell: _,
+                env,
+                working_directory: _,
+            } => match env {
+                LoE::Expr(e) => {
+                    Self::check_secrets_access(e.as_bare(), step, &mut findings)?;
```

Comment:
I think this check needs to be a little more discerning: `e` needs to be parsed into an `Expr`, which then needs to be walked recursively to visit each `Expr::Context` and match _that_ on the `secrets` namespace. 

Without that, we'll end up with (admittedly goofy) false positives like:

```yaml
with:
  yolo: ${{ 'secrets' }}
```

Writing the AST walker is currently a bit annoying/manual; if you want to wait until I add a `Visitor` API that might make your life a bit easier (I plan on doing that pretty soon) :slightly_smiling_face:

Reply:
I like that! It will also increase the audit confidence. I am ok with waiting for the `Visitor` API, just give me a ping once ready and I'll put it to work :)